File: api/ViaQuestInc.StepOne.Core/Candidates/Workflows/CandidateWorkflowStep.cs
using System.Text.Json.Serialization;
using ViaQuestInc.StepOne.Core.Data.Entity;
using ViaQuestInc.StepOne.Core.Workflows.Steps;

namespace ViaQuestInc.StepOne.Core.Candidates.Workflows;

public class CandidateWorkflowStep : EntityBase<int>
{
    public required int CandidateWorkflowId { get; set; }

    [JsonIgnore]
    public CandidateWorkflow? CandidateWorkflow { get; set; }

    public required int WorkflowStepId { get; set; }

    public WorkflowStep? WorkflowStep { get; set; }

    public DateTime? CompletedAt { get; set; }

    public bool IsCompleted => CandidateWorkflowStepStatus == CandidateWorkflowStepStatuses.Completed;

    public bool IsConfirmedByAdmin { get; set; }

    public CandidateWorkflowStepStatuses CandidateWorkflowStepStatus { get; set; }
}

File: ui/src/app/models/workflows/candidate-workflow.step.ts
import {CandidateWorkflow} from "./candidate.workflow";
import {WorkflowStep} from "./workflow.step";
import {CandidateWorkflowStepStatuses} from "./candidate-workflow-step.statuses";

/**
 * Represents a step within a candidate's workflow.
 */
export interface CandidateWorkflowStep {
  id: number;

  /** The ID of the associated candidate workflow. */
  candidateWorkflowId: number;

  /** The candidate workflow associated with this step. */
  candidateWorkflow?: CandidateWorkflow | null;

  /** The ID of the associated workflow step. */
  workflowStepId: number;

  /** The workflow step associated with this candidate workflow step. */
  workflowStep?: WorkflowStep | null;

  /** The timestamp when this step was completed, if applicable. */
  completedAt?: Date | null;

  /** Whether this step has been completed. */
  isCompleted: boolean;

  /** Indicates whether an administrator has confirmed this step. */
  isConfirmedByAdmin: boolean;

  /** The current status of this candidate workflow step. */
  candidateWorkflowStepStatus: CandidateWorkflowStepStatuses;
}
